File: lib/features/chat/domain/utils/chat_settings.dart
import 'dart:convert';

/// Inner rumor kind used for 1:1 chat settings signaling (Iris convention).
///
/// See `nostr-double-ratchet` v0.0.59 "Disappearing Message Signaling".
const int kChatSettingsKind = 10448;

/// Parsed "chat-settings" payload.
class ChatSettingsPayload {
  const ChatSettingsPayload({required this.messageTtlSeconds});

  /// Per-chat disappearing message timer in seconds.
  ///
  /// `null` means "Off".
  final int? messageTtlSeconds;
}

/// Parse a "chat-settings" JSON payload from a settings rumor's `content`.
///
/// Expected shape:
/// `{ "type": "chat-settings", "v": 1, "messageTtlSeconds": <seconds|null> }`
ChatSettingsPayload? parseChatSettingsContent(String content) {
  if (content.isEmpty) return null;

  Object? decoded;
  try {
    decoded = jsonDecode(content);
  } catch (_) {
    return null;
  }
  if (decoded is! Map<String, dynamic>) return null;

  if (decoded['type'] != 'chat-settings') return null;
  if (decoded['v'] != 1) return null;

  final ttl = decoded['messageTtlSeconds'];
  if (ttl == null) {
    return const ChatSettingsPayload(messageTtlSeconds: null);
  }
  if (ttl is num) {
    final normalized = ttl.floor();
    return ChatSettingsPayload(
      messageTtlSeconds: normalized > 0 ? normalized : null,
    );
  }

  return null;
}

/// Build a "chat-settings" payload JSON for sending.
String buildChatSettingsContent({required int? messageTtlSeconds}) {
  final normalized = (messageTtlSeconds != null && messageTtlSeconds > 0)
      ? messageTtlSeconds
      : null;
  return jsonEncode(<String, Object?>{
    'type': 'chat-settings',
    'v': 1,
    'messageTtlSeconds': normalized,
  });
}

File: src/lib/chatSettings.ts
import { normalizeDisappearingTtl } from './disappearingNotice'

export interface ChatSettingsPayloadV1 {
  type: 'chat-settings'
  v: 1
  messageTtlSeconds: number | null
}

export function parseChatSettingsContent(content: string): ChatSettingsPayloadV1 | null {
  if (!content) return null
  let parsed: unknown
  try {
    parsed = JSON.parse(content)
  } catch {
    return null
  }

  if (!parsed || typeof parsed !== 'object') return null
  const obj = parsed as Record<string, unknown>
  if (obj.type !== 'chat-settings') return null
  if (obj.v !== 1) return null

  const ttl = obj.messageTtlSeconds
  if (ttl === null) {
    return { type: 'chat-settings', v: 1, messageTtlSeconds: null }
  }
  if (typeof ttl === 'number' && Number.isFinite(ttl)) {
    return {
      type: 'chat-settings',
      v: 1,
      messageTtlSeconds: normalizeDisappearingTtl(ttl),
    }
  }

  return null
}
